utils/s3/client.cc

```diff
@@ -234,26 +235,92 @@ future<> map_s3_client_exception(std::exception_ptr ex) {
+
+    co_return aws::aws_error{aws::aws_error_type::OK, aws::retryable::no};
+}
 
+future<> client::do_retryable_request(group_client& gc, http::request req, http::experimental::client::reply_handler handler) {
+    return do_with(uint32_t{0}, std::move(req), std::move(handler),
+            [this, &gc](uint32_t& retries, http::request& request, http::experimental::client::reply_handler& handler) -> future<> {
+                return repeat_until_value([&gc, &retries, &request, &handler, this]() -> future<std::optional<uint32_t>> {
+                    if (retries <= _retry_strategy->get_max_retries()) {
+                        return gc.http.make_request(request, handler, std::nullopt)
+                                .then([&retries] {
+                                    return make_ready_future<std::optional<uint32_t>>(retries);
+                                })
+                                .handle_exception_type([this, &retries](const aws::aws_exception& ex) mutable -> future<std::optional<uint32_t>> {
+                                    ++retries;
+                                    if (_retry_strategy->should_retry(ex.error(), retries)) {
+                                        s3l.debug("S3 client request failed. Reason: {}. Retry# {}", ex.what(), retries);
+                                        co_await seastar::sleep(std::chrono::milliseconds(_retry_strategy->delay_before_retry(ex.error(), retries)));
+                                    } else {
+                                        if (!ex.error().is_retryable()) {
+                                            s3l.error("S3 client encountered non-retryable error. Reason: {}. Code: {}. Retry# {}", ex.what(),
+                                                    std::to_underlying(ex.error().get_error_type()), retries);
+                                        } else {
+                                            s3l.error("Retries exhausted. Retry# {}", retries);
+                                        }
+                                        co_return coroutine::return_exception(ex);
+                                    }
+                                    co_return std::optional<uint32_t>(std::nullopt);
+                                });
+                    }
+                    return make_exception_future<std::optional<uint32_t>>(std::runtime_error("Retries exhausted"));
+                }).then([](auto) {
+                    return make_ready_future<>();
+                });
+            })
+            .handle_exception([](auto ex) {
+                return map_s3_client_exception(std::move(ex));
+            });
+}
+
+static http::experimental::client::reply_handler make_error_aware_handler(
+        const http::request& req, http::experimental::client::reply_handler&& handler, http::reply::status_type expected) {
+    return [is_mpu_completion{req._method == "POST" && !req.get_query_param("uploadId").empty()}, handler{std::move(handler)}, expected](
```

Comment:
what does "mpu" mean? would be nice if you could either expand it to something like "multipart_upload" ? and, if the caller understands if it is performing a multiple part upload, how about accepting a parameter to this function, like
```c++
using multipart_upload = bool_class<class multi_cell_tag>;
static http::experimental::client::reply_handler make_error_aware_handler(
        const http::request& req,
        http::experimental::client::reply_handler&& handler,
        multipart_upload is_multipart_upload,
        http::reply::status_type expected) {
  // ...
}
```

Reply:
ok, passed an argument all the way from the `finalize_upload`